drivers/timer/apic_tsc.c

```diff
@@ -174,6 +182,23 @@ static int sys_clock_driver_init(void)
 	__ASSERT((ebx & BIT(1)) != 0, "No TSC_ADJUST MSR support");
 #endif
 
+#ifdef CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME
+	int tsc_freq;
+
+	eax = 0;
+	ebx = 0;
+	ecx = 0;
+	if (!__get_cpuid(0x15, &eax, &ebx, &ecx, &edx)) {
+		__ASSERT(false, "error cpuid leaf 0x15\n");
+	}
+
+	__ASSERT((ebx != 0) && (ecx != 0), "No TSC/core crystal clock ratio is enumerated");
+
+	tsc_freq = ecx * ebx/eax;
```

Comment:
These are 32 bit values, and the precedence here is going to compute (ecx * ebx) / eax.  Does the architecture guarantee that the numerator can't overflow?  Probably best to do the multiply in 64 bit.

Reply:
Current Zephyr supported Intel platform limited to maximum 2.4GHz. The z_clock_hw_cycles_per_sec is also 32bit variable currently

Reply:
The **result** is representable in 32 bits.  But the intermediate product computed before the division may not be.  Add a cast and do it in 64 bits.

Reply:
updated